src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs

```diff
@@ -273,80 +274,149 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
                 _tracer.Log($"Ovewriting existing File {bundlePath}");
             }
 
-            BinaryUtils.CopyFile(hostSource, bundlePath);
-
-            // Note: We're comparing file paths both on the OS we're running on as well as on the target OS for the app
-            // We can't really make assumptions about the file systems (even on Linux there can be case insensitive file systems
-            // and vice versa for Windows). So it's safer to do case sensitive comparison everywhere.
-            var relativePathToSpec = new Dictionary<string, FileSpec>(StringComparer.Ordinal);
-
             long headerOffset = 0;
-            using (FileStream bundle = File.Open(bundlePath, FileMode.Open, FileAccess.ReadWrite))
-            using (BinaryWriter writer = new BinaryWriter(bundle, Encoding.Default, leaveOpen: true))
+            string destinationDirectory = new FileInfo(bundlePath).Directory?.FullName ?? "";
+            if (!Directory.Exists(destinationDirectory))
             {
-                if (_target.IsOSX)
+                Directory.CreateDirectory(destinationDirectory);
+            }
+            using (FileStream bundleStream = File.Open(bundlePath, FileMode.Create, FileAccess.ReadWrite))
+            {
+                using (FileStream hostSourceStream = File.OpenRead(hostSource))
                 {
-                    MachObjectFile.RemoveCodeSignatureIfPresent(bundle);
+                    hostSourceStream.CopyTo(bundleStream);
                 }
-                bundle.Position = bundle.Length;
-                foreach (var fileSpec in fileSpecs)
+                EmbeddedSignatureBlob? signatureBlob = null;
+                if (_target.IsOSX)
                 {
-                    string relativePath = fileSpec.BundleRelativePath;
-
-                    if (IsHost(relativePath))
+                    long? newLength = null;
+                    using (MemoryMappedFile bundleMap = MemoryMappedFile.CreateFromFile(bundleStream, null, bundleStream.Length, MemoryMappedFileAccess.ReadWrite, HandleInheritability.None, leaveOpen: true))
+                    using (MemoryMappedViewAccessor viewAccessor = bundleMap.CreateViewAccessor())
                     {
-                        continue;
+                        MachObjectFile file = MachObjectFile.Create(viewAccessor);
+                        signatureBlob = file.EmbeddedSignatureBlob;
+                        file.RemoveCodeSignatureIfPresent(viewAccessor, out newLength);
                     }
-
-                    if (ShouldIgnore(relativePath))
+                    if (newLength.HasValue)
                     {
-                        _tracer.Log($"Ignore: {relativePath}");
-                        continue;
+                        bundleStream.SetLength(newLength.Value);
                     }
+                }
+                using (BinaryWriter writer = new BinaryWriter(bundleStream, Encoding.Default, leaveOpen: true))
```

Comment:
Could we also do this with memory-mapping, rather than re-opening the written file?

Reply:
Yeah, we probably could calculate the size of the files ahead of time and extend the file to that length before memory mapping.

Reply:
Updated to use a single memory-mapped file rather than memory mapping multiple times.